[AF 1140] Poms Migration tool

desmax74 opened the pull request:
@porcelli @jakubschwan feedbacks are welcome

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomEditor.java:
```diff
@@ -0,0 +1,414 @@
+    private boolean write(Model model, String absolutePath){
+
+        try {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
```
You're not closing this stream. Please use `try-with-resources` to guarantee stream is closed automatically
```
try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
    ...
}
```

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomEditor.java:
```diff
@@ -0,0 +1,414 @@
+    protected String FILE_URI = "file://";
+    private static final String KIE_VERSION_KEY = "KIE_VERSION";
+    private String kieVersion;
+    private String SCOPE_PROVIDED = "provided";
```
How about making all of these final (or don't use Uppercase names fi they're not constants) to avoid confusion?

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomEditor.java:
```diff
@@ -0,0 +1,414 @@
+
+    private List<Dependency> applyMandatoryDeps(List<Dependency> dependencies) {
+        Set<Dependency> uniques = new HashSet<>();
+        uniques.addAll(dependencies);
```
You don't need to call addAll - you can pass the original collection directly to constructor

```
Set<Dependency> uniques = new HashSet<>(dependencies);
```

desmax74 replied:
@jhrcek a explanation to avoid a waste of energy, there are a lot of changes because the origin of the branchis the maven compiler branch, the code and files to review are in the module kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration, you could skip the files in others module. Thanks

jhrcek replied:
Ok, I'll focus on that module only.

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomEditor.java:
```diff
@@ -49,82 +50,82 @@
+    public final String TRUE = "true";
     private final Logger logger = LoggerFactory.getLogger(PomEditor.class);
+    //protected String FILE_URI = "file://";
+    protected String FILE_URI = "file:" + File.separator + File.separator;
```
This one is probably wrong - file URL should always have "//" , regardless of what OS we are on.
`protected String FILE_URI = "file://";` was correct.

jhrcek commented on kie-wb-common-ala/kie-wb-common-ala-wildfly/kie-wb-common-ala-wildfly-provider/pom.xml:
```diff
@@ -156,7 +156,7 @@
             <execution>
               <id>ban-duplicated-classes</id>
-              <phase>validate</phase>
+              <phase>accept</phase>
```
@desmax74 Even though this is outside of the scope of the review, this should be probably reverted. It looks like you did some renaming refactoring in your IDE (renaming validate to accept), but this renamed all textual occurences of validate -just search for `accept` in this PR and you'll see many invalid replacements in comments etc.

desmax74 replied:
I'll align with the master to fix all this "spurious" changes

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PluginPresence.java:
```diff
@@ -0,0 +1,35 @@
+public class PluginPresence {
+
+    private boolean isPresent;
+    private int position = -1;
```
You can just remove this `= -1` because this field is always set by the constructor

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomEditor.java:
```diff
@@ -0,0 +1,400 @@
+            updateRepositories(model);
+            updatePluginRepositories(model);
+            boolean writed = write(model, pom.toAbsolutePath().toString());
+            if (writed) {
```
typo writed -> written

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/test/java/org/kie/workbench/common/project/migration/cli/maven/PomProcessorTest.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Test
+    @Ignore
```
Please add info about why this is ignored (something based on which someone can decide about when to unignore it, otherwise it will be ignored forever) or remove the annotation.

`@Ignore("Failing because ABC. Can be unignored after XYZ is fixed")`

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/test/java/org/kie/workbench/common/project/migration/cli/maven/PomProcessorTest.java:
```diff
@@ -0,0 +1,59 @@
+        FileUtils.copyDirectory(path.toFile(), pathCopy.toFile());
+        try {
+            int pomProcessed = processor.processProject(path.toAbsolutePath().toString());
+            assertTrue(pomProcessed == 1);
```
`asserEquals(1, pomProcessed)` is better

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/test/java/org/kie/workbench/common/project/migration/cli/maven/PomProcessorTest.java:
```diff
@@ -0,0 +1,59 @@
+            assertTrue(pomProcessed == 1);
+        } catch (Exception e) {
+            fail(e.getMessage());
+        } finally {
```
Please use `@After` for cleanup instead of finally.

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomEditor.java:
```diff
@@ -0,0 +1,400 @@
+        } else {
+            try {
+                prop.load(in);
+                in.close();
```
Please close this stream in finally :)

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomEditor.java:
```diff
@@ -0,0 +1,400 @@
+    private boolean write(Model model, String absolutePath) {
+
+        try {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
```
You're not closing this stream. Please use try-with-resources to guarantee this stream is closed
```
try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
   ...
}
```

desmax74 commented:
@jhrcek now is more clean and simple to review, applied the previous suggestions :)

desmax74 commented:
@porcelli  Alex checks if this meet all new requirements about the pom changes

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/pom.xml:
```diff
@@ -0,0 +1,141 @@
+    </dependency>
+    <!--
+
+    <dependency>
```
Let's not commit commented-out configuration, please.

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/PromptPomMigrationService.java:
```diff
@@ -0,0 +1,84 @@
+
+    public String promptForExternalConfiguration() {
+        SystemAccess.Console console = system.console();
+        console.format("WARNING: Please ensure that you have made backups of the directory [%s] before proceeding.\n", niogitDir);
```
Do we need 2nd prompt for backup here? It's already in `promptForBackup()`..

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomEditor.java:
```diff
@@ -0,0 +1,301 @@
+                return new Model();
+            }
+        } catch (Exception e) {
+            System.out.println("Error occurred during POMs migration:" + e.getMessage());
```
We should probably use `SystemAccess` or logging here instead of `System.out.println`. The same on line 104

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomJsonReader.java:
```diff
@@ -0,0 +1,177 @@
+        try {
+            JsonReader reader = Json.createReader(in);
+            pomObject = reader.readObject();
+            reader.close();
```
Please close in finally, or even better use try-with-resources which does that automatically for anything implementing `AutoCloseable`.

```
try (JsonReader reader = Json.createReader(in)) {
    pomObject = reader.readObject();
} catch (Exception e) {
    logger.error(e.getMessage());
    throw new RuntimeException(e.getMessage());
 }
```

desmax74 commented:
@jhrcek @jakubschwan is it ok ?

porcelli commented:
@jhrcek this is for GA, so we need get it merged sooner than later.
@krisv what are your thoughts about this timeline?

krisv commented:
@porcelli @jhrcek yes, ideally we would like to see this merged this week but I realize that might be very difficult.  Our code freeze is next Wed so we would really need to be merged by then in 7.7.x or it would miss the CR1 cut.

jhrcek commented:
Ok, I'll test this locally today and provide feedback asap.

desmax74 commented:
@jhrcek about the blocker and major issues

1) If a git root repo is used a barerepo is copied and no pom.xml are founded in the tmpdir @porcelli in which branch of the https://github.com/desmax74/kie-wb-common/blob/AF-1140/kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-api/src/main/java/org/kie/workbench/common/migration/cli/MigrationApp.java#L40 the PomMigration is mandatory ? 

2) Fixed

3) I could fix adding a logback.xml in the kie-wb-common-cli-api but in the ER5 
RHPAM-7.0.0.ER5 isn't present and the debug msgs aren't present

4) Fixed

5) Fixed

jhrcek commented:
@desmax74 regarding issue 3: When you run migration in ER5, you'll see this in the output:

```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
```

Which means that ER5 doesn't have logging implementation (confirmed that by looking into migration-tool/lib/).

Your PR adds `logback-classic `and `logback-core` which ends up in the zip, which leads to those debug messages being printed. I think that adding reasonable logback.xml is a good idea!

jhrcek commented:
@desmax74 So I tested this locally and the migrated pom.xml are written to the migrated repositories :+1: 

The bad thing though is that the tool still produces crazy amounts of logging output. There's no more  DEBUG output, but there are still tons of info and warning messages - see example output 
[migrationOutput2.txt](https://github.com/kiegroup/kie-wb-common/files/1979052/migrationOutput2.txt)

The main problem with this is that at the end of migration there's a ton of exception messages (probably result of stopping weld container), in which the tool prompts are completely lost (so I can't see things like "Press 4 to exit").

I think we should disable console logging (so that only things logged via `SystemAccess` would be visible) and only redirect logging output to some log file (so customers can potentially send this file for debugging) - wdyt?

desmax74 commented:
@jhrcek The log messages could be reduced with different log level. The exceptions are the output of the automatic builds failed due to the dependencies without version because aren't our dependencies (like internal rh or qe deps), one solution is stop the build on updates on  files or disable the output and write on a log file like you are proposing, if we prevent the automatic build the users could see this problems on the pom only when they run a build, if we not prevent the automatic build they reads this problems on a file later (with a correct appender log configuration) wdyt ? And about the path of the log file, which path we have to use ?

desmax74 commented:
@jhrcek changed in the following way, the console shows only the list of the projects at the end the message "End POMs migration, detailed log available on <user.home>/pom_migration.log", the errors are contained in the file log

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-api/src/main/java/org/kie/workbench/common/migration/cli/MigrationSetup.java:
```diff
@@ -28,5 +28,6 @@ public static void configureProperties(final SystemAccess system,
         system.setProperty(JGitFileSystemProviderConfiguration.GIT_NIO_DIR_NAME, niogitDir.getFileName().toString());
         system.setProperty(JGitFileSystemProviderConfiguration.GIT_DAEMON_ENABLED, "false");
         system.setProperty(JGitFileSystemProviderConfiguration.GIT_SSH_ENABLED, "false");
+        //system.setProperty("org.uberfire.watcher.autostart", "false");//uncommnet to disabled the automatic build on update files
```
+1 to uncomment this line - in my opinion it makes sense to disable automatic build after migration.

desmax74 replied:
@jhrcek if this line is uncommented the watching service is stopped and when a pom is changed the automatic build won't start, the result is the migration_tool.log will be without errors, the user will see the errors due to lack of version in the user dependencies only when a build will starts. Is this the desidered behaviour ?

jhrcek replied:
I'm not sure. Let's see what others think about this setting.. @mbarkley ?

pefernan replied:
+1 to uncomment this, not sure what side effects could appears, but I think it would be ok.

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/PomMigrationTool.java:
```diff
@@ -0,0 +1,161 @@
+                    }
+                }
+                String userHome = System.getProperty("user.home");
+                system.out().println("End POMs migration, detailed log available on "+userHome + "/migration_tool.log");
```
"End POMs migration" -> Finished POMs migration

jhrcek replied:
log available on -> log available in

desmax74 replied:
Ok

jhrcek commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/PomMigrationTool.java:
```diff
@@ -0,0 +1,161 @@
+                        processWorkspaceProject(prj, jsonPath);
+                    }
+                }
+                String userHome = System.getProperty("user.home");
```
In my opinion using "user.dir" (the dir from which user started the tool) makes more sense.

desmax74 replied:
Ok

desmax74 commented:
jenkins retest this

desmax74 commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/ServiceCDIWrapper.java:
```diff
@@ -0,0 +1,65 @@
+import org.uberfire.io.IOService;
+
+@ApplicationScoped
+public class ServiceCDIWrapper {
```
Ok, where do you suggest to place ?

porcelli replied:
if shared is needed, maybe create a module cli-shared

desmax74 replied:
Is fine the cli-api name, here the common parts of the MigrationServicesCDIWrapper will be moved to the cli-api, then the FormDefinitionSerializer will be moved from MigrationServiceCDIWrapper into FormsMigrationServiceCDIWrapper in the cli-forms-migration because it needs classes from the forms migration module

mbarkley commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomJsonReader.java:
```diff
@@ -0,0 +1,190 @@
+
+/**
+ * It reads the pom-migrations.json with dependencies, repositories and plugin repositories with the following format
+ * <p>
```
Having the example is good, but it isn't readable as is. I think if you use the `<pre></pre>` tags, the line-wrapping will be preserved in the javadocs.

desmax74 replied:
Correct

mbarkley commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomJsonReader.java:
```diff
@@ -0,0 +1,190 @@
+public class PomJsonReader {
+
+    private final Logger logger = LoggerFactory.getLogger(PomJsonReader.class);
+    private String JSON_POM_FILE;//= "pom-migration.json";
```
Since this isn't a constant, could you change it to camel case?

desmax74 replied:
Ok

mbarkley commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomJsonReader.java:
```diff
@@ -0,0 +1,190 @@
+        } catch (Exception e) {
+            logger.error(e.getMessage());
+            throw new RuntimeException(e.getMessage());
+        }finally {
```
Some of the spacing is off in this block. Maybe run the formatter?

desmax74 replied:
ok

mbarkley commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomJsonReader.java:
```diff
@@ -0,0 +1,190 @@
+        try (JsonReader reader = Json.createReader(in)) {
+            pomObject = reader.readObject();
+        } catch (Exception e) {
+            logger.error(e.getMessage());
```
You probably want to add the exception as the second argument to keep the stacktrace. Also, in the line below it would be good to add the caught exception as a cause to the thrown exception.

desmax74 replied:
ok

mbarkley commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/test/java/org/kie/workbench/common/project/migration/cli/maven/PomEditorTest.java:
```diff
@@ -0,0 +1,188 @@
+            }
+        } catch (Exception e) {
+            logger.error(e.getMessage());
+            fail(e.getMessage());
```
Definitely we'll want to see the stacktrace if this test ever fails. The pattern I'd recommend for this is to manually throw an  `AssertionError` instead of calling `fail`. Then you can add `e` as a cause.

desmax74 replied:
sounds like a GOTO but is ok in a test

mbarkley commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/test/java/org/kie/workbench/common/project/migration/cli/maven/PomEditorTest.java:
```diff
@@ -0,0 +1,188 @@
+            assertTrue(modelUpdated.getRepositories().size() == 2);
+            assertTrue(modelUpdated.getPluginRepositories().size() == 2);
+        } catch (Exception e) {
+            logger.error(e.getMessage());
```
Same note on showing stracktrace. See comment above.

desmax74 replied:
changed in the same way

mbarkley commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/test/java/org/kie/workbench/common/project/migration/cli/maven/PomEditorTest.java:
```diff
@@ -0,0 +1,188 @@
+            modelUpdated = editor.updatePom(null, null);
+            assertTrue(modelUpdated.getGroupId() == null);
+        } catch (Exception e) {
+            logger.error(e.getMessage());
```
Same comment as above.

desmax74 replied:
changed in the same way

desmax74 commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/PomMigrationTool.java:
```diff
@@ -0,0 +1,161 @@
+
+    @Override
+    public String getDescription() {
+        return "Migrates pom.xml files to format required in 7.7.x and later (community)";
```
I agree

desmax74 commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/PomMigrationTool.java:
```diff
@@ -0,0 +1,161 @@
+                return FileVisitResult.CONTINUE;
+            }
+        });
+        system.out().println("Migrated " + counter[0] + " POMs for prj:"+workspaceProject.getName());
```
Ok

desmax74 commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/PomMigrationTool.java:
```diff
@@ -0,0 +1,161 @@
+                                counter[0]++;
+                            }
+                        } catch (Exception e) {
+                            system.err().println("Error reading form: " + fileName + ":\n");
```
ok

desmax74 commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/DependencyKey.java:
```diff
@@ -0,0 +1,51 @@
+    private Dependency dependency;
+
+    public DependencyKey(Dependency dep) {
+        this.dependency = dep;
```
ok

desmax74 commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomEditor.java:
```diff
@@ -0,0 +1,343 @@
+
+    private void updateRepositories(Model model) {
+        List<Repository> repos = model.getRepositories();
+        if (repos == null || repos.isEmpty()) {
```
Correct isn't needed

desmax74 commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomEditor.java:
```diff
@@ -0,0 +1,343 @@
+
+    private void updatePluginRepositories(Model model) {
+        List<Repository> repos = model.getPluginRepositories();
+        if (repos == null || repos.isEmpty()) {
```
Correct isn't needed

desmax74 commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomEditor.java:
```diff
@@ -0,0 +1,343 @@
+                             "Pom's Migration" + path.toString());
+            return true;
+        } catch (Exception e) {
+            logger.error(e.getMessage());
```
ok

desmax74 commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomEditor.java:
```diff
@@ -0,0 +1,343 @@
+        }
+    }
+
+    private boolean writeToFile(Model model, String absolutePath) {
```
+1

desmax74 commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/PromptPomMigrationService.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private boolean jsonExists(String path) {
+        File test = new File(path + File.separator + JSON_POM_FILE);
```
Just replaced, thanks for the tip :)

desmax74 commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/PomMigrationTool.java:
```diff
@@ -0,0 +1,161 @@
+        if (projectMigrationWasExecuted()) {
+            try {
+                MigrationSetup.configureProperties(system, niogitDir);
+                weldContainer = new Weld().initialize();
```
Ok

desmax74 commented:
@mbarkley changes applied

porcelli commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-api/src/main/java/org/kie/workbench/common/migration/cli/MigrationSetup.java:
```diff
@@ -28,5 +28,6 @@ public static void configureProperties(final SystemAccess system,
         system.setProperty(JGitFileSystemProviderConfiguration.GIT_NIO_DIR_NAME, niogitDir.getFileName().toString());
         system.setProperty(JGitFileSystemProviderConfiguration.GIT_DAEMON_ENABLED, "false");
         system.setProperty(JGitFileSystemProviderConfiguration.GIT_SSH_ENABLED, "false");
+        system.setProperty("org.uberfire.watcher.autostart", "false");//commen2t to enable the automatic build on update files
```
the comment `//commen2t to enable the automatic build on update files` could be removed

desmax74 replied:
ok

porcelli commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-forms-migration/src/main/java/org/kie/workbench/common/forms/migration/tool/cdi/MigrationServicesCDIWrapper.java:
```diff
@@ -40,6 +40,8 @@
 
     private IOService systemIoService;
 
+    //@TODO move in shared place to reuse in all migration tools
```
todo? why not move?

desmax74 replied:
Because it requires works on the Pere's module

porcelli replied:
check with pere, and move it to a common module.

porcelli commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/PomMigrationTool.java:
```diff
@@ -0,0 +1,154 @@
+
+    private SystemAccess system;
+    private ToolConfig config;
+    private String POM_DOT_XML = "pom.xml";
```
static final?

desmax74 replied:
only final, is here to avoid a "pom.xml" on a if check

porcelli commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/PromptPomMigrationService.java:
```diff
@@ -0,0 +1,63 @@
+
+    public String promptForExternalConfiguration() {
+        SystemAccess.Console console = system.console();
+        Collection<String> validResponses = Arrays.asList("yes", "no");
```
instead of "yes" "no" strings, please use Enum

desmax74 replied:
Is the pattern used in the other modules

porcelli commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomEditor.java:
```diff
@@ -0,0 +1,311 @@
+    private static final String JSON_POM_MANDATORY_DEPS = "pom-mandatory.json";
+    private static final String KIE_VERSION_KEY = "KIE_VERSION";
+    private final Logger logger = LoggerFactory.getLogger(PomEditor.class);
+    private String FILE_URI = "file://";
```
static final

desmax74 replied:
removed wasn't used

porcelli commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PomEditor.java:
```diff
@@ -0,0 +1,311 @@
+    private String KJAR_PKG = "kjar";
+    private String POM_PKG = "pom";
+    private String KIE_MAVEN_PLUGIN_ARTIFACT_ID = "kie-maven-plugin";
+    private JSONDTO jsonConf;
```
all upper cased should be static final

desmax74 replied:
ok

porcelli commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/test/java/org/kie/workbench/common/project/migration/cli/maven/PomEditorTest.java:
```diff
@@ -0,0 +1,194 @@
+    private String EVALUATION_PRJ = "/target/test-classes/employee-rostering/";
+    private String ITORDERS_PRJ = "/target/test-classes/itorders/";
+    private String MORTGAGES_PRJ = "/target/test-classes/mortgages/";
+    private String OPTACLOUD_PRJ = "/target/test-classes/optacloud/";
```
uppercase == static final

desmax74 replied:
ok

porcelli commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/PromptPomMigrationService.java:
```diff
@@ -0,0 +1,63 @@
+        Collection<String> validResponses = Arrays.asList("yes", "no");
+        String response;
+        do {
+            response = console.readLine("Do you want provide a path to an external Pom migration json  with dependencies and repositories ? [yes/no]: ").toLowerCase();
```
extra whitespace

desmax74 replied:
fixed

porcelli commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/DependencyKey.java:
```diff
@@ -0,0 +1,61 @@
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
```
null check?

desmax74 replied:
ok

mbarkley replied:
See comment in equals method of dependency.

porcelli commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/DependenciesCollection.java:
```diff
@@ -0,0 +1,50 @@
+
+import org.apache.maven.model.Dependency;
+
+public class DependenciesCollection {
```
Pojo missing toString, equals and hashCode

desmax74 replied:
ok

porcelli commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/JSONDTO.java:
```diff
@@ -0,0 +1,46 @@
+import org.apache.maven.model.Dependency;
+import org.apache.maven.model.Repository;
+
+public class JSONDTO {
```
missing toString, equals and hashCode

desmax74 replied:
ok

porcelli commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PluginPresence.java:
```diff
@@ -0,0 +1,35 @@
+ */
+package org.kie.workbench.common.project.migration.cli.maven;
+
+public class PluginPresence {
```
toString, equals and hashCode

desmax74 replied:
ok

mbarkley commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/DependenciesCollection.java:
```diff
@@ -47,4 +49,30 @@ public void addDependenciesKeys(List<DependencyKey> deps) {
+
+    @Override
+    public boolean equals(Object o) {
+        PortablePreconditions.checkNotNull("DependenciesCollection", o);
```
Hmmm, this will throw an exception if compared to `null`, which I think is not ideal. I hate to contradict @porcelli, but I think it was okay the way it was before because `instanceof` is always false for null objects, so a null argument would cause the method to exit on line 68.

@porcelli WDYT?

porcelli replied:
@mbarkley is correct that the use of checkNotNull is not ideal. When I suggested null check was more like what my IDE usually generates:

```java
        if (o == null) {
            return false;
        }
        if (this == o) {
            return true;
        }
        if (!(o instanceof ???)) {
            return false;
        }
        ??? that = (???) o;
        ....
```

mbarkley commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/JSONDTO.java:
```diff
@@ -43,4 +45,35 @@ public JSONDTO(List<Dependency> dependencies, List<Repository> respositories, Li
+
+    @Override
+    public boolean equals(Object o) {
+        PortablePreconditions.checkNotNull("JSONDTO", o);
```
Same comment as above.

mbarkley commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/java/org/kie/workbench/common/project/migration/cli/maven/PluginPresence.java:
```diff
@@ -32,4 +36,33 @@ public boolean isPresent() {
+
+    @Override
+    public boolean equals(Object o) {
+        PortablePreconditions.checkNotNull("PluginPresence", o);
```
Same comment as above.

mbarkley commented on kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/test/java/org/kie/workbench/common/project/migration/cli/maven/PomEditorTest.java:
```diff
@@ -85,9 +87,13 @@ private void testDefault(String prj) {
-            fail(e.getMessage());
+        } catch (IOException ioex) {
+            logger.error(ioex.getMessage(), ioex);
+            throw new AssertionError(ioex.getMessage());
```
What I had in mind here, was this:

     throw new AssertionError("Test failed from error.", ioex);

That way when the test fails you will get the error stacktrace as part of the failure. Same applies to other AssertionErrors below.

desmax74 commented:
jenkins retest this

porcelli commented:
@desmax74 found a small bug. existing dependencies are loosing the version, example:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId>guvnor.feature</groupId>
  <artifactId>dtables</artifactId>
  <version>1.0</version>
  <packaging>kjar</packaging>
  <name>dtables</name>
  <dependencies>
    <dependency>
      <groupId>org.group</groupId>
      <artifactId>workitems</artifactId>
      <version>1.0.0-FINAL</version>
    </dependency>
  </dependencies>
</project>
```

results in this:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId>guvnor.feature</groupId>
  <artifactId>dtables</artifactId>
  <version>1.0</version>
  <packaging>kjar</packaging>
  <name>dtables</name>
  <dependencies>
    <dependency>
      <groupId>com.thoughtworks.xstream</groupId>
      <artifactId>xstream</artifactId>
      <version>1.4.10</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.kie</groupId>
      <artifactId>kie-internal</artifactId>
      <version>7.8.0-SNAPSHOT</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>org.optaplanner</groupId>
      <artifactId>optaplanner-core</artifactId>
      <version>7.8.0-SNAPSHOT</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>4.12</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.group</groupId>
      <artifactId>workitems</artifactId>
    </dependency>
    <dependency>
      <groupId>org.optaplanner</groupId>
      <artifactId>optaplanner-persistence-jaxb</artifactId>
      <version>7.8.0-SNAPSHOT</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>org.kie</groupId>
      <artifactId>kie-api</artifactId>
      <version>7.8.0-SNAPSHOT</version>
      <scope>provided</scope>
    </dependency>
  </dependencies>
  <build>
    <plugins>
      <plugin>
        <groupId>org.kie</groupId>
        <artifactId>kie-maven-plugin</artifactId>
        <version>7.8.0-SNAPSHOT</version>
        <extensions>true</extensions>
      </plugin>
    </plugins>
  </build>
</project>
```

porcelli commented on kie-wb-common-services/kie-wb-common-services-backend/src/main/resources/pom-mandatory.json:
```diff
@@ -0,0 +1,10 @@
+    {"groupId":"junit", "artifactId":"junit", "version":"4.12", "scope":"test"},
+    {"groupId":"com.thoughtworks.xstream", "artifactId":"xstream", "version":"1.4.10", "scope":"test"}
+  ]
+}
```
this file needs to be filtered, this is what I get when I run it:

```java
<?xml version="1.0" encoding="UTF-8"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.myteam</groupId>
  <artifactId>MyTest</artifactId>
  <version>1.0.0</version>
  <packaging>kjar</packaging>
  <name>MyTest</name>
  <description></description>
  <dependencies>
    <dependency>
      <groupId>com.thoughtworks.xstream</groupId>
      <artifactId>xstream</artifactId>
      <version>1.4.10</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.kie</groupId>
      <artifactId>kie-internal</artifactId>
      <version>${version.org.kie}</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>org.optaplanner</groupId>
      <artifactId>optaplanner-core</artifactId>
      <version>${version.org.kie}</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>4.12</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.optaplanner</groupId>
      <artifactId>optaplanner-persistence-jaxb</artifactId>
      <version>${version.org.kie}</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>org.kie</groupId>
      <artifactId>kie-api</artifactId>
      <version>${version.org.kie}</version>
      <scope>provided</scope>
    </dependency>
  </dependencies>
  <repositories>
    <repository>
      <id>guvnor-m2-repo</id>
      <name>Guvnor M2 Repo</name>
      <url>file:///Users/porcelli/Desktop/wildfly-11.0.0.Final/repositories/kie/global</url>
    </repository>
  </repositories>
  <build>
    <plugins>
      <plugin>
        <groupId>org.kie</groupId>
        <artifactId>kie-maven-plugin</artifactId>
        <version>${version.org.kie}</version>
        <extensions>true</extensions>
      </plugin>
    </plugins>
  </build>
</project>
```

desmax74 replied:
@porcelli the version.org.kie was retrieved with https://github.com/desmax74/kie-wb-common/blob/AF-1140/kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-pom-migration/src/main/resources/PomMigration.properties and was filtered in the cli-pom-migration. Added the filtering true in the kie-wb-common-services-backend to here  filter https://github.com/desmax74/kie-wb-common/tree/AF-1140/kie-wb-common-services/kie-wb-common-services-backend/src/main/resources

desmax74 replied:
Test added
